File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.cpp
class Solution {
public:
    // we execute this soln in such a way , that after incrementing i we make it equal to the size
  /*  bool solve(vector<int> &nums,int i,vector<int> dp){
        if(i == nums.size()) return true;
        if(i > nums.size()) return false;
        if(dp[i] != -1) return dp[i];
        
        if(i+1<nums.size() && nums[i]==nums[i+1]){
            // 2 elements are equal
            if(solve(nums,i+2,dp)) return true;
            // 3 elements are equal
            if(i+2<nums.size() && nums[i+1]==nums[i+2]){
                if(solve(nums,i+3,dp)) return true;
            }
        }
        // if nums[i]+2=nums[i+1]+1=nums[i]
        if(i+2 < nums.size() && nums[i]==nums[i+1]-1 && nums[i]==nums[i+2]-2)
            if(solve(nums,i+3,dp)) return true;
        // if none of the cases are satisfied then return false
        return dp[i]=false;
    }*/
    bool solve(vector<int>& nums, int i, vector<int>& dp){
        if(i == nums.size()) return true;
        if(i > nums.size()) return false;
        if(dp[i] != -1) return dp[i];
        
        if(i + 1 < nums.size() && nums[i] == nums[i+1] ) {
            if(solve(nums, i + 2, dp)) return true;
            if(i + 2 < nums.size() && nums[i] == nums[i+2]){
                if(solve(nums, i + 3, dp)) return true;
            }
        }
        if(i + 2 < nums.size() && nums[i] == nums[i+1]-1 && nums[i] == nums[i+2] - 2 ){
            if(solve(nums, i + 3, dp)) return true;
        }
        return dp[i] = false;
    }
    bool validPartition(vector<int>& nums) {
        // if size is 2 we check for this and return
        if(nums.size()==2) return nums[0]==nums[1];
        //declare a 1d dp & init to -1
        vector<int> dp(nums.size(),-1);
        // solve the dp by memoisation method
        return solve(nums,0,dp);
    }
};

File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.py
class Solution:
    def validPartition(self, nums: List[int]) -> bool:
        n=len(nums)
        @cache
        def recur(i):
            
            if i>=n: 
                # print(i)
                return True
            o1=o2=o3=False
            
            
            if i+1<n and i+2<n and nums[i]==nums[i+1]==nums[i+2]:
                # print(i+3,i)
                o1=recur(i+3)
            if i+1<n and i+2<n and nums[i]+2==nums[i+1]+1==nums[i+2]:
                # print(i+3,i)
                o2=recur(i+3)
            if i+1<n and nums[i]==nums[i+1]:
                # print(i+2,i)
                o3=recur(i+2)
            if o1 or o2 or o3:
                return True
            else:
                return False
            
        return recur(0)